cabal-install/Distribution/Client/Outdated.hs

```diff
@@ -0,0 +1,134 @@
+import Distribution.Client.Setup
+import Distribution.Client.Targets
+import Distribution.Client.Types
+import Distribution.Solver.Types.PackageConstraint
+import Distribution.Solver.Types.PackageIndex
+import Distribution.Client.Sandbox.PackageEnvironment
+
+import Distribution.Package                          (PackageName, packageVersion)
+import Distribution.PackageDescription               (buildDepends)
+import Distribution.PackageDescription.Configuration (finalizePD)
+import Distribution.PackageDescription.Parse
+       (readPackageDescription)
+import Distribution.Simple.Compiler                  (Compiler, compilerInfo)
+import Distribution.Simple.Setup                     (fromFlagOrDefault)
+import Distribution.Simple.Utils
+       (die, notice, debug, tryFindPackageDesc)
+import Distribution.System                           (Platform)
+import Distribution.Text                             (display)
+import Distribution.Types.ComponentRequestedSpec
+       (defaultComponentRequestedSpec)
+import Distribution.Types.Dependency
+       (Dependency(..), depPkgName, simplifyDependency)
+import Distribution.Verbosity                        (Verbosity)
+import Distribution.Version
+       (Version, UpperBound(..), asVersionIntervals, majorBoundVersion)
+
+import qualified Data.Set as S
+import System.Directory                              (getCurrentDirectory)
+import System.Exit                                   (exitFailure)
+
+-- | Entry point for the 'outdated' command.
+outdated :: Verbosity -> OutdatedFlags -> RepoContext
+         -> Compiler -> Platform
+         -> IO ()
+outdated verbosity outdatedFlags repoContext comp platform = do
+  sourcePkgDb <- IndexUtils.getSourcePackages verbosity repoContext
+  let freeze   = fromFlagOrDefault False (outdatedFreeze outdatedFlags)
+      exitCode = fromFlagOrDefault False (outdatedExitCode outdatedFlags)
+      ignore   = S.fromList (outdatedIgnore outdatedFlags)
+      minor    = S.fromList (outdatedMinor outdatedFlags)
+      pkgIndex = packageIndex sourcePkgDb
+  deps <- if freeze
+          then depsFromFreezeFile verbosity
+          else depsFromPkgDesc    verbosity comp platform
+  let outdatedDeps = listOutdated deps pkgIndex ignore minor
+  notice verbosity ("Outdated dependencies: "
+                     ++ intercalate ", "
+                        (map (\(d, v) -> display d
+                               ++ " (latest: " ++ display v ++ ")") outdatedDeps))
+  if (exitCode && (not . null $ outdatedDeps))
+    then exitFailure
+    else return ()
+
+-- | Read the list of dependencies from the freeze file.
+depsFromFreezeFile :: Verbosity -> IO [Dependency]
+depsFromFreezeFile verbosity = do
+  cwd        <- getCurrentDirectory
+  userConfig <- loadUserConfig verbosity cwd Nothing
+  let ucnstrs = map fst . configExConstraints . savedConfigureExFlags $ userConfig
+      deps    = mapMaybe (packageConstraintToDependency . userToPackageConstraint)
+                ucnstrs
```

Comment:
This code is simply converting the list of `UserConstraint`s from the freeze file to `[Dependency]` so that we can look them up in the package index. So it doesn't rely on them being in any special form.